hermes/bin/src/logger.rs

```diff
@@ -0,0 +1,145 @@
+//! Setup for logging for the service.
+
+use std::str::FromStr;
+
+use clap::ValueEnum;
+extern crate derive_more;
+use derive_more::Display;
+use tracing::{level_filters::LevelFilter, subscriber::SetGlobalDefaultError};
+use tracing_subscriber::{
+    fmt::{format::FmtSpan, time},
+    FmtSubscriber,
+};
+
+use crate::runtime_extensions::bindings::hermes::logging::api::Level;
+
+/// All valid logging levels.
+#[derive(ValueEnum, Clone, Copy, Display, Default)]
+pub(crate) enum LogLevel {
+    /// Errors
+    #[display(fmt = "Error")]
+    Error,
+    /// Warnings
+    #[display(fmt = "Warn")]
+    Warn,
+    /// Informational Messages
+    #[display(fmt = "Info")]
+    #[default]
+    Info,
+    /// Debug messages
+    #[display(fmt = "Debug")]
+    Debug,
+    /// Tracing
+    #[display(fmt = "Trace")]
+    Trace,
+}
+
+impl FromStr for LogLevel {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // Error and Warn levels are force to Info level
+        // as Info is the highest log level one can choose.
+        match s {
+            "error" | "warn" | "info" => Ok(LogLevel::Info),
+            "debug" => Ok(LogLevel::Debug),
+            "trace" => Ok(LogLevel::Trace),
+            _ => Err(()),
+        }
+    }
+}
+
+impl From<Level> for LogLevel {
+    fn from(level: Level) -> Self {
+        // Error and Warn levels are force to Info level
+        // as Info is the highest log level one can choose.
+        match level {
+            Level::Info | Level::Warn | Level::Error => LogLevel::Info,
+            Level::Debug => LogLevel::Debug,
+            Level::Trace => LogLevel::Trace,
+        }
+    }
+}
+
+/// Implements a conversion from `LogLevel` enum to the `tracing::Level`.
+impl From<LogLevel> for tracing::Level {
+    fn from(val: LogLevel) -> Self {
+        match val {
+            LogLevel::Error => Self::ERROR,
+            LogLevel::Warn => Self::WARN,
+            LogLevel::Info => Self::INFO,
+            LogLevel::Debug => Self::DEBUG,
+            LogLevel::Trace => Self::TRACE,
+        }
+    }
+}
+
+/// Logger configuration.
+#[derive(Default)]
+pub(crate) struct LoggerConfig {
+    /// Log level
+    log_level: LogLevel,
+    /// Enable/disable thread logging
+    with_thread: bool,
+    /// Enable/disable file logging
+    with_file: bool,
+    /// Enable/disable line number logging
+    with_line_num: bool,
+}
+
+#[allow(dead_code)]
+impl LoggerConfig {
```

Comment:
If you want to implement a builder pattern, you should split this logic to the few structs `LoggerConfigBuilder` and 
`LoggerConfig`.
So `LoggerConfigBuilder` would have what you have already written for `LoggerConfig` instead of `build` function will return `LoggerConfig`